File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    void dfs(int row, int col, vector<vector<char>>& grid, vector<vector<bool>>& visited) {
        int rows = grid.size();
        int cols = grid[0].size();

        if(row < 0 || row >= rows || col < 0 || col >= cols)
            return;
        if(grid[row][col] == '0' || visited[row][col]) 
            return;

        visited[row][col] = true;
        dfs(row + 1, col, grid, visited);
        dfs(row - 1, col, grid, visited);
        dfs(row, col - 1, grid, visited);
        dfs(row, col + 1, grid, visited);
    }

    int numIslands(vector<vector<char>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();

        vector<vector<bool>>visited(rows, vector<bool>(cols, false));
        int islandCount = 0;

        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                if(grid[row][col] == '1' && !visited[row][col]) {
                    islandCount++;
                    dfs(row, col, grid, visited);
                }
            }
        }
        return islandCount;
    }
};

File: 200-number-of-islands/number-of-islands.java
class Solution {
    public int rows, cols;

    public void DFS(int i, int j, char[][] grid, boolean[][] visited) {
        if(i < 0 || i >= rows || j < 0 || j >= cols) 
            return;

        if(grid[i][j] == '0' || visited[i][j] == true) 
            return;
        
        visited[i][j] = true;

        DFS(i + 1, j, grid, visited);
        DFS(i - 1, j, grid, visited);
        DFS(i, j + 1, grid, visited);
        DFS(i, j - 1, grid, visited);
    }

    public int numIslands(char[][] grid) {
        this.rows = grid.length;
        this.cols = grid[0].length;
        int landCount = 0;

        boolean[][] visited = new boolean[rows][cols];
        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(grid[i][j] == '1' && visited[i][j] == false) {
                    DFS(i, j, grid, visited);
                    landCount++;
                }
            }
        }
        return landCount;
    }
}